File: api-sample/oneec-csharp-sample/Oneec-Sample/services/OneECAPI.cs
using Newtonsoft.Json;
using NPOI.POIFS.Crypt;
using Oneec_Sample.models;
using System;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Web;

namespace Oneec_Sample.services

{
    class OneECAPI
    {
        // dev url
        const string _endpoint = "https://dev-api.oneec.ai";

        /* Please contact the official counterpart */
        // partner key id
        string _partnerKeyId = "";
        // aes key 
        string _AESKey = "";
        // aes iv
        string _AESIV = "";
        // hash key
        string _hashKey = "";
        // merchant account token
        string _merchantAccountToken = "";
        public OneECAPI()
        {
        }

        private void SetHeader(HttpClient httpClient, string apiRoute, string body)
        {
            var authorization = $"{_partnerKeyId}.{_merchantAccountToken}";
            var target = apiRoute + body + _hashKey;
            var xSign = CryptService.SHA256Hash(target);
            httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {authorization}");
            httpClient.DefaultRequestHeaders.Add("X-sign", xSign);
            httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }

        /// <summary>
        /// Get orders
        /// </summary>
        public void GetOrderList()
        {
            try
            {
                HttpClient httpClient = new HttpClient();
                string apiRoute = "/oapi/v1/data/merchant/orders";

                string param = "";
                //param = param+"?";
                //param = param + "&shipStartDate=2022-03-16T16:46:05.005Z";      // ship start date  default "" 
                //param = param + "&shipEndDate=2023-03-16T16:46:05.005Z";        // ship end date  default "" 
                //param = param + "&orderCreateDate=2022-03-16T16:46:05.005Z";    // order create date  default ""  
                //param = param + "&orderStatus=";                                // 
                //param = param + "&channelId=AyNAVo";                            // channel id
                //param = param + "&channelSettingId=partner_unit_test";          // platform
                //param = param + "&start=0";                                     // start default 0
                //param = param + "&limit=20";                                    // page size default 10
                string body = "";
                var endpoint = $"{_endpoint}{apiRoute}{param}";
                SetHeader(httpClient, apiRoute + param, body);
                var message = httpClient.GetAsync(endpoint).Result;
                string response = message.Content.ReadAsStringAsync().Result;
                Console.WriteLine(response);

                var objectModel = JsonConvert.DeserializeObject<Response<String>>(response);
                if (objectModel.status == 200)
                {
                    var data = CryptService.AesGcmDecryptTByBase64(_AESKey, _AESIV, objectModel.data);
                    Console.WriteLine(data);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }

        }

        /// <summary>
        /// Get the products
        /// </summary>
        public void GetProducts()
        {
            try
            {
                HttpClient httpClient = new HttpClient();
                string apiRoute = "/oapi/v1/data/merchant/products";
                string param = "";
                //param = param+"?";
                //param = param + "&limit=2";      // page size default 10   ,   max=100
                //param = param + "&start=9";      // start default 0
                //param = param + "&itemNumber=";  // product's number
                string body = "";
                var endpoint = $"{_endpoint}{apiRoute}{param}";
                SetHeader(httpClient, apiRoute + param, body);

                var message = httpClient.GetAsync(endpoint).Result;
                string resultStr = message.Content.ReadAsStringAsync().Result;
                Console.WriteLine(resultStr);
                var objectModel = JsonConvert.DeserializeObject<Response<MerchantProductData>>(resultStr);

            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }

        /// <summary>
        /// Get the combination products
        /// </summary>
        public void GetMerchantCombinationProductList()
        {
            try
            {
                HttpClient httpClient = new HttpClient();
                string apiRoute = "/oapi/v1/data/merchant/product_combinations";
                string param = "";
                //param = param+"?";
                //param = param + "&limit=2";      // page size default 10   ,   max=100
                //param = param + "&start=9";      // start default 0
                //param = param + "&itemNumber=";  // product's number
                string body = "";
                var endpoint = $"{_endpoint}{apiRoute}{param}";
                SetHeader(httpClient, apiRoute+ param, body);

                var message = httpClient.GetAsync(endpoint).Result;
                string resultStr = message.Content.ReadAsStringAsync().Result;
                Console.WriteLine(resultStr);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }

        }

        /// <summary>
        /// Create or modify product
        /// </summary>
        public void SaveProduct()
        {
            try
            {
                HttpClient httpClient = new HttpClient();
                string apiRoute = "/oapi/v1/data/merchant/products/save";
                var specs = new MerchantProductSpec[1];
                specs[0] = new MerchantProductSpec() { };
                var payload = new MerchantProduct
                {
                    productName = "TBSProduct",     
                    brand = "The Body Shop 美體小舖",                    
                    supplier = "美雅國際",              
                    condition = 0,
                    goodType = 1,
                    distributionTemperature = 0,
                    isFragile = false,
                    specs = specs,
                    suggestPrice = 500,
                    normalPrice = 400,
                    cost = 300,
                    qty = 50,
                    isCanOvertakeOrder = true,
                    itemNumber = "TBS12345",
                    insertDt = "2022-05-05T22:23:11.333Z",
                    modifiedDt = "2022-05-05T22:23:11.333Z",
                    size = new MerchantProductSize() { height = 1, length = 1, weight = 2, width = 5 },
                };
                var body = JsonConvert.SerializeObject(payload);
                var endpoint = $"{_endpoint}{apiRoute}";
                SetHeader(httpClient, apiRoute, body);

                Console.WriteLine(body);
                var httpContent = new StringContent(body, Encoding.UTF8, "application/json");
                var message = httpClient.PostAsync(endpoint, httpContent).Result;
                string resultStr = message.Content.ReadAsStringAsync().Result;
                Console.WriteLine(resultStr);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }

        }

        /// <summary>
        /// Create or modify combination products
        /// </summary>
        public void SaveProductCombinations()
        {
            try
            {
                HttpClient httpClient = new HttpClient();
                string apiRoute = "/oapi/v1/data/merchant/product_combinations/save";
                var data = new MerchantCombinationProduct();
                data.itemNumber = "1234533";
                data.addCombinationInfo(new MerchantCombinationInfo() { itemNumber = "070105022", cost = 33, price = 3, productName = "TBS-A", qty = 1 });
                data.addCombinationInfo(new MerchantCombinationInfo() { itemNumber = "0696204", cost = 333, price = 6, productName = "TBS-B", qty = 2 });
                data.modifiedDt = "2022-05-05T22:23:11.333Z";
                data.insertDt = "2022-05-05T22:23:11.333Z";
                var body = JsonConvert.SerializeObject(data);
                var endpoint = $"{_endpoint}{apiRoute}";
                SetHeader(httpClient, apiRoute, body);

                Console.WriteLine(body);
                var httpContent = new StringContent(body, Encoding.UTF8, "application/json");
                var message = httpClient.PostAsync(endpoint, httpContent).Result;
                string resultStr = message.Content.ReadAsStringAsync().Result;
                Console.WriteLine(resultStr);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }
        
        public void GetChannels()
        {
            try
            {
                HttpClient httpClient = new HttpClient();
                string apiRoute = "/oapi/v1/data/merchant/channels/actived";
                string param = "";
                Console.WriteLine(param);
                string body = "";
                var endpoint = $"{_endpoint}{apiRoute}{param}";
                SetHeader(httpClient, apiRoute + param, body);
                var message = httpClient.GetAsync(endpoint).Result;
                string response = message.Content.ReadAsStringAsync().Result;
                Console.WriteLine(response);

            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }
    }
}

File: api-sample/oneec-java-sample/src/main/java/oneec/java/sample/OneECAPI.java
package oneec.java.sample;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.json.JsonMapper;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.List;

import oneec.java.sample.helper.CryptHelper;
import oneec.java.sample.models.Response;
import oneec.java.sample.models.CipherResponse;
import oneec.java.sample.models.MerchantCombinationProduct;
import oneec.java.sample.models.MerchantProduct;

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

public class OneECAPI {
    private static final String _domain = "https://dev-api.oneec.ai"; // Dev url
    private static final String _partnerKeyID = ""; // Please contact the official counterpart
    private static final String _aesKey = ""; // Please contact the official counterpart
    private static final String _aesIv = ""; // Please contact the official counterpart
    private static final String _hashKey = ""; // Please contact the official counterpart
    private static final String _merchantAccessToken = ""; // Please contact the official counterpart
    private static CryptHelper _cryptHelper;

    private static String getXSign(String apiUrl, String body) throws Exception {
        String sign = apiUrl + body + _hashKey;

        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        digest.update(sign.getBytes("utf8"));
        return String.format("%064x", new BigInteger(1, digest.digest()));
    }

    public static void main(String[] args) {
        _cryptHelper = new CryptHelper(_aesKey, _aesIv, "AES/GCM/NoPadding");

        try {
            getOrderList();
            getProductList();
            getMerchantCombinationProductList();
            saveProduct();
            saveProductCombinations();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * Get the order list
     */
    public static void getOrderList() {
        try {
            String apiRoute = "/oapi/v1/data/merchant/orders";

            String param = "";
            //param = param+"?";
            //param = param + "&shipStartDate=2022-03-16T16:46:05.005Z";      // ship start date  default ""
            //param = param + "&shipEndDate=2023-03-16T16:46:05.005Z";        // ship end date  default ""
            //param = param + "&orderCreateDate=2022-03-16T16:46:05.005Z";    // order create date  default ""
            //param = param + "&orderStatus=";                                //
            //param = param + "&channelId=AyNAVo";                            // channel id
            //param = param + "&channelSettingId=partner_unit_test";          // platform
            //param = param + "&start=0";                                     // start default 0
            //param = param + "&limit=20";                                    // page size default 10

            String endpoint = _domain + apiRoute + param;

            String body = "";

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(_partnerKeyID + "." + _merchantAccessToken);
            headers.add("X-sign", getXSign(apiRoute + param, body));

            HttpEntity<String> request = new HttpEntity<>(body, headers);

            ResponseEntity<String> entity = new RestTemplate().exchange(endpoint, HttpMethod.GET, request, String.class);

            TypeReference<CipherResponse> typeReference = new TypeReference<>() {
            };

            CipherResponse response = new JsonMapper().readValue(entity.getBody(), typeReference);
            if (response.status == HttpStatus.OK.value()) {
                String decryptedData = _cryptHelper.decryptFromBase64(response.data);
                System.out.println("decryptedData: " + decryptedData);
            } else {
                System.out.println("response: " + new JsonMapper().writeValueAsString(response));
            }
        } catch (Exception e) {
            System.out.println(e.getMessage() + ", " + e);
        }
    }

    /**
     * Get the product list
     */
    public static void getProductList() {
        try {
            String apiRoute = "/oapi/v1/data/merchant/products";

            String param = "";
            //param = param+"?";
            //param = param + "&limit=2";      // page size default 10
            //param = param + "&start=9";      // start default 0
            //param = param + "&itemNumber=";  // product's number

            String endpoint = _domain + apiRoute + param;

            String body = "";

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(_partnerKeyID + "." + _merchantAccessToken);
            headers.add("X-sign", getXSign(apiRoute + param, body));

            HttpEntity<String> request = new HttpEntity<>(body, headers);

            ResponseEntity<String> entity = new RestTemplate().exchange(endpoint, HttpMethod.GET, request, String.class);

            TypeReference<Response> typeReference = new TypeReference<>() {
            };

            Response response = new JsonMapper().readValue(entity.getBody(), typeReference);
            System.out.println(response.data);

            if (response.status == HttpStatus.OK.value()) {
                System.out.println("data: " + response.data);
            } else {
                System.out.println("response: " + new JsonMapper().writeValueAsString(response));
            }
        } catch (Exception e) {
            System.out.println(e.getMessage() + ", " + e);
        }
    }

    /**
     * Get the combination product list
     */
    public static void getMerchantCombinationProductList() {
        try {
            String apiRoute = "/oapi/v1/data/merchant/product_combinations";

            String endpoint = _domain + apiRoute;

            String param = "";
            //param = param+"?";
            //param = param + "&limit=2";      // page size default 10
            //param = param + "&start=9";      // start default 0
            //param = param + "&itemNumber=";  // product's number

            String body = "";

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(_partnerKeyID + "." + _merchantAccessToken);
            headers.add("X-sign", getXSign(apiRoute + param, body));

            HttpEntity<String> request = new HttpEntity<>(body, headers);

            ResponseEntity<String> entity = new RestTemplate().exchange(endpoint, HttpMethod.GET, request, String.class);

            TypeReference<Response> typeReference = new TypeReference<>() {
            };

            Response response = new JsonMapper().readValue(entity.getBody(), typeReference);
            System.out.println(response.data);

            if (response.status == HttpStatus.OK.value()) {
                System.out.println("data: " + response.data);
            } else {
                System.out.println("response: " + new JsonMapper().writeValueAsString(response));
            }
        } catch (Exception e) {
            System.out.println(e.getMessage() + ", " + e);
        }
    }

    /**
     * Create or modify product
     */
    public static void saveProduct() {
        try {
            String apiRoute = "/oapi/v1/data/merchant/products/save";

            String endpoint = _domain + apiRoute;

            MerchantProduct merchantProduct = new MerchantProduct();
            merchantProduct.productName = "TBSProduct";
            merchantProduct.brand = "The Body Shop 美體小舖";
            merchantProduct.supplier = "美雅國際";
            merchantProduct.condition = 0;
            merchantProduct.goodType = 1;
            merchantProduct.distributionTemperature = 0;
            merchantProduct.isFragile = false;
            merchantProduct.specs = new ArrayList<>();
            merchantProduct.suggestPrice = BigDecimal.valueOf(500);
            merchantProduct.normalPrice = BigDecimal.valueOf(400);
            merchantProduct.cost = BigDecimal.valueOf(300);
            merchantProduct.qty = 50;
            merchantProduct.isCanOvertakeOrder = true;
            merchantProduct.itemNumber = "TBS12345";
            merchantProduct.insertDt = "2022-05-05T22:23:11.333Z";
            merchantProduct.modifiedDt = "2022-05-05T22:23:11.333Z";

            MerchantProduct.MerchantProductSize size = new MerchantProduct.MerchantProductSize();
            size.height = 1;
            size.length = 1;
            size.weight = 1;
            size.width = 1;

            merchantProduct.size = size;
            merchantProduct.specs = new ArrayList<>();

            String body = new JsonMapper().writeValueAsString(merchantProduct);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(_partnerKeyID + "." + _merchantAccessToken);
            headers.add("X-sign", getXSign(apiRoute, body));

            HttpEntity<String> request = new HttpEntity<>(body, headers);

            ResponseEntity<String> entity = new RestTemplate().exchange(endpoint, HttpMethod.POST, request, String.class);

            TypeReference<Response> typeReference = new TypeReference<>() {
            };

            Response response = new JsonMapper().readValue(entity.getBody(), typeReference);
            System.out.println(response.data);

            if (response.status == HttpStatus.OK.value()) {
                System.out.println("data: " + response.data);
            } else {
                System.out.println("response: " + new JsonMapper().writeValueAsString(response));
            }
        } catch (Exception e) {
            System.out.println(e.getMessage() + ", " + e);
        }
    }

    /**
     * Create or modify combination products
     */
    public static void saveProductCombinations() {
        try {
            String apiRoute = "/oapi/v1/data/merchant/product_combinations/save";

            String endpoint = _domain + apiRoute;

            MerchantCombinationProduct merchantCombinationProduct = new MerchantCombinationProduct();
            merchantCombinationProduct.itemNumber = "1234533";
            merchantCombinationProduct.modifiedDt = "2022-05-12T14:52:11.333Z";
            merchantCombinationProduct.insertDt = "2022-05-12T14:52:11.333Z";

            List<MerchantCombinationProduct.MerchantCombinationInfo> combinationInfoList = new ArrayList<>();

            MerchantCombinationProduct.MerchantCombinationInfo combinationInfoI01 = new MerchantCombinationProduct.MerchantCombinationInfo();
            combinationInfoI01.itemNumber = "070105022";
            combinationInfoI01.productName = "TBS-A";
            combinationInfoI01.qty = 1;
            combinationInfoI01.cost = BigDecimal.valueOf(33);
            combinationInfoI01.price = BigDecimal.valueOf(3);
            combinationInfoList.add(combinationInfoI01);

            MerchantCombinationProduct.MerchantCombinationInfo combinationInfoI02 = new MerchantCombinationProduct.MerchantCombinationInfo();
            combinationInfoI02.itemNumber = "0696204";
            combinationInfoI02.productName = "TBS-B";
            combinationInfoI02.qty = 2;
            combinationInfoI02.cost = BigDecimal.valueOf(333);
            combinationInfoI02.price = BigDecimal.valueOf(6);
            combinationInfoList.add(combinationInfoI02);
            merchantCombinationProduct.combinationInfos = combinationInfoList;

            String body = new JsonMapper().writeValueAsString(merchantCombinationProduct);
            System.out.println(body);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(_partnerKeyID + "." + _merchantAccessToken);
            headers.add("X-sign", getXSign(apiRoute, body));

            HttpEntity<String> request = new HttpEntity<>(body, headers);

            ResponseEntity<String> entity = new RestTemplate().exchange(endpoint, HttpMethod.POST, request, String.class);

            TypeReference<Response> typeReference = new TypeReference<>() {
            };

            Response response = new JsonMapper().readValue(entity.getBody(), typeReference);
            System.out.println(response.data);

            if (response.status == HttpStatus.OK.value()) {
                System.out.println("data: " + response.data);
            } else {
                System.out.println("response: " + new JsonMapper().writeValueAsString(response));
            }
        } catch (Exception e) {
            System.out.println(e.getMessage() + ", " + e);
        }

    }

}
